core/index/index_writer.cpp

```diff
@@ -2555,196 +2571,169 @@ index_writer::pending_context_t index_writer::flush_all(
+}
 
-    meta_.segments_ = to_commit.meta->segments_;  // create copy
+bool IndexWriter::Start(ProgressReportCallback const& progress) {
+  IRS_ASSERT(!commit_lock_.try_lock());  // already locked
 
-    // 1st phase of the transaction successfully finished here,
-    // set to_commit as active flush context containing pending meta
-    pending_state_.commit = std::make_shared<committed_state_t::element_type>(
-      std::piecewise_construct,
-      std::forward_as_tuple(std::move(to_commit.meta)),
-      std::forward_as_tuple(std::move(pending_refs)));
-  } catch (...) {
-    writer_->rollback();  // rollback started transaction
+  REGISTER_TIMER_DETAILED();
 
-    throw;
+  if (pending_state_.Valid()) {
+    // begin has been already called
+    // without corresponding call to commit
+    return false;
   }
 
-  // only noexcept operations below
+  auto to_commit = FlushAll(progress);
 
-  // release cached readers
-  cached_readers_.purge(to_commit.ctx->segment_mask_);
-  pending_state_.ctx = std::move(to_commit.ctx);
+  if (to_commit.Empty()) {
+    // nothing to commit, no transaction started
+    return false;
+  }
+
+  StartImpl(std::move(to_commit.ctx), {.index_meta = std::move(to_commit.meta)},
+            std::move(to_commit.files_to_sync));
 
   return true;
 }
 
-void index_writer::finish() {
+void IndexWriter::Finish() {
   IRS_ASSERT(!commit_lock_.try_lock());  // already locked
 
   REGISTER_TIMER_DETAILED();
 
-  if (!pending_state_) {
+  if (!pending_state_.Valid()) {
     return;
   }
 
-  Finally reset_state = [this]() noexcept {
-    // release reference to flush_context
-    pending_state_.reset();
+  bool res{false};
+
+  Finally reset_state = [this, &res]() noexcept {
+    if (IRS_UNLIKELY(!res)) {
+      Abort();  // Rollback failed transaction
+    }
+
+    // Release reference to flush_context
+    pending_state_.Reset();
   };
 
-  // lightweight 2nd phase of the transaction
+  // Lightweight 2nd phase of the transaction
 
-  try {
-    if (!writer_->commit()) {
-      throw illegal_state{"Failed to commit index metadata."};
-    }
-#ifndef __APPLE__
-    // atomic_store may throw
-    static_assert(!noexcept(
-      std::atomic_store(&committed_state_, std::move(pending_state_.commit))));
-#endif
-    std::atomic_store(&committed_state_, std::move(pending_state_.commit));
-  } catch (...) {
-    abort();  // rollback transaction
+  res = writer_->commit();
 
-    throw;
+  if (IRS_UNLIKELY(!res)) {
+    throw illegal_state{"Failed to commit index metadata."};
   }
 
-  // after here transaction successfull (only noexcept operations below)
-
-  // update 'last_gen_' to last commited/valid generation
-  meta_.last_gen_ = committed_state_->first->gen_;
+  // after this line transaction is successfull (only noexcept operations below)
+  std::atomic_store_explicit(&committed_reader_,
+                             std::move(pending_state_.commit),
+                             std::memory_order_relaxed);
```

Comment:
Do you sure about relaxed?
It looks like common acquire/release pattern:
Because we construct data which when will be stored to `committed_reader_` variable